src/contrasts.jl

```diff
@@ -567,20 +614,94 @@ function contrasts_matrix(C::ContrastsCoding, baseind, n)
 end
 
 ## hypothesis matrix
-needs_intercept(mat) =
+"""
+    needs_intercept(mat::AbstractMatrix)
+
+Returns `true` if a contrasts matrix `mat` needs to consider the intercept to
+determine the corresponding hypothesis matrix.  This is determined by whether
+`rank(mat)` is less than the number of rows and any column is not orthogonal to
+the intercept column.
+"""
+needs_intercept(mat::AbstractMatrix) =
     (rank(mat) < size(mat, 1)) &&
-    !all(mapslices(sum, mat, dims=1) .< 10eps(eltype(mat)))
+    !all(isapprox.(sum(mat, dims=1), 0, rtol=1e-5))
 
-hypothesis_matrix(contrasts::AbstractContrasts, baseind, n; kwargs...) =
-    hypothesis_matrix(contrasts_matrix(contrasts, baseind, n); kwargs...)
+"""
+    hypothesis_matrix(cmat::AbstractMatrix; intercept=needs_intercept(cm), pretty=true)
+    hypothesis_matrix(contrasts::AbstractContrasts, n; baseind=1, kwargs...)
+    hypothesis_matrix(cmat::ContrastsMatrix; kwargs...)
+
+Compute the hypotehsis matrix for a contrasts matrix using the generalized
+pseudo-inverse (`LinearAlgebra.pinv`).  `intercept` determines whether a column
+of ones is included before taking the pseudoinverse, which is needed for
+contrasts where the columns are not orthogonal to the intercept (e.g., have
+non-zero mean).  If `pretty=true` (the default), the hypotheses are rounded to
```

Comment:
Instead of `pretty`, this could be a `tolerance` argument defaulting to `0`?

Reply:
good idea

Reply:
actually, I'd rather have a nonzero tolerance, to default to pretty-printing the matrix.  and setting it to 0 would opt out of rationalizing it.